File: app/src/test/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApiTest.java
package fr.free.nrw.commons.mwapi;

import android.os.Build;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import fr.free.nrw.commons.BuildConfig;
import io.reactivex.observers.TestObserver;
import okhttp3.HttpUrl;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 21)
public class ApacheHttpClientMediaWikiApiTest {

    private ApacheHttpClientMediaWikiApi testObject;
    private MockWebServer server;

    @Before
    public void setUp() throws Exception {
        server = new MockWebServer();
        testObject = new ApacheHttpClientMediaWikiApi("http://" + server.getHostName() + ":" + server.getPort() + "/");
        testObject.setWikiMediaToolforgeUrl("http://" + server.getHostName() + ":" + server.getPort() + "/");
    }

    @After
    public void teardown() throws IOException {
        server.shutdown();
    }

    @Test
    public void authCookiesAreHandled() {
        assertEquals("", testObject.getAuthCookie());

        testObject.setAuthCookie("cookie=chocolate-chip");

        assertEquals("cookie=chocolate-chip", testObject.getAuthCookie());
    }

    @Test
    public void simpleLoginWithWrongPassword() throws Exception {
        server.enqueue(new MockResponse().setBody("<?xml version=\"1.0\"?><api batchcomplete=\"\"><query><tokens logintoken=\"baz\" /></query></api>"));
        server.enqueue(new MockResponse().setBody("<?xml version=\"1.0\"?><api><clientlogin status=\"FAIL\" message=\"Incorrect password entered.&#10;Please try again.\" messagecode=\"wrongpassword\" /></api>"));

        String result = testObject.login("foo", "bar");

        RecordedRequest loginTokenRequest = assertBasicRequestParameters(server, "POST");
        Map<String, String> body = parseBody(loginTokenRequest.getBody().readUtf8());
        assertEquals("xml", body.get("format"));
        assertEquals("query", body.get("action"));
        assertEquals("login", body.get("type"));
        assertEquals("tokens", body.get("meta"));

        RecordedRequest loginRequest = assertBasicRequestParameters(server, "POST");
        body = parseBody(loginRequest.getBody().readUtf8());
        assertEquals("1", body.get("rememberMe"));
        assertEquals("foo", body.get("username"));
        assertEquals("bar", body.get("password"));
        assertEquals("baz", body.get("logintoken"));
        assertEquals("https://commons.wikimedia.org", body.get("loginreturnurl"));
        assertEquals("xml", body.get("format"));

        assertEquals("wrongpassword", result);
    }

    @Test
    public void simpleLogin() throws Exception {
        server.enqueue(new MockResponse().setBody("<?xml version=\"1.0\"?><api batchcomplete=\"\"><query><tokens logintoken=\"baz\" /></query></api>"));
        server.enqueue(new MockResponse().setBody("<?xml version=\"1.0\"?><api><clientlogin status=\"PASS\" username=\"foo\" /></api>"));

        String result = testObject.login("foo", "bar");

        RecordedRequest loginTokenRequest = assertBasicRequestParameters(server, "POST");
        Map<String, String> body = parseBody(loginTokenRequest.getBody().readUtf8());
        assertEquals("xml", body.get("format"));
        assertEquals("query", body.get("action"));
        assertEquals("login", body.get("type"));
        assertEquals("tokens", body.get("meta"));

        RecordedRequest loginRequest = assertBasicRequestParameters(server, "POST");
        body = parseBody(loginRequest.getBody().readUtf8());
        assertEquals("1", body.get("rememberMe"));
        assertEquals("foo", body.get("username"));
        assertEquals("bar", body.get("password"));
        assertEquals("baz", body.get("logintoken"));
        assertEquals("https://commons.wikimedia.org", body.get("loginreturnurl"));
        assertEquals("xml", body.get("format"));

        assertEquals("PASS", result);
    }

    @Test
    public void twoFactorLogin() throws Exception {
        server.enqueue(new MockResponse().setBody("<?xml version=\"1.0\"?><api batchcomplete=\"\"><query><tokens logintoken=\"baz\" /></query></api>"));
        server.enqueue(new MockResponse().setBody("<?xml version=\"1.0\"?><api><clientlogin status=\"PASS\" username=\"foo\" /></api>"));

        String result = testObject.login("foo", "bar", "2fa");

        RecordedRequest loginTokenRequest = assertBasicRequestParameters(server, "POST");
        Map<String, String> body = parseBody(loginTokenRequest.getBody().readUtf8());
        assertEquals("xml", body.get("format"));
        assertEquals("query", body.get("action"));
        assertEquals("login", body.get("type"));
        assertEquals("tokens", body.get("meta"));

        RecordedRequest loginRequest = assertBasicRequestParameters(server, "POST");
        body = parseBody(loginRequest.getBody().readUtf8());
        assertEquals("1", body.get("rememberMe"));
        assertEquals("foo", body.get("username"));
        assertEquals("bar", body.get("password"));
        assertEquals("baz", body.get("logintoken"));
        assertEquals("1", body.get("logincontinue"));
        assertEquals("2fa", body.get("OATHToken"));
        assertEquals("xml", body.get("format"));

        assertEquals("PASS", result);
    }

    @Test
    public void validateLoginForLoggedInUser() throws Exception {
        server.enqueue(new MockResponse().setBody("<?xml version=\"1.0\"?><api><query><userinfo id=\"10\" name=\"foo\"/></query></api>"));

        boolean result = testObject.validateLogin();

        RecordedRequest loginTokenRequest = assertBasicRequestParameters(server, "GET");
        Map<String, String> body = parseQueryParams(loginTokenRequest);
        assertEquals("xml", body.get("format"));
        assertEquals("query", body.get("action"));
        assertEquals("userinfo", body.get("meta"));

        assertTrue(result);
    }

    @Test
    public void validateLoginForLoggedOutUser() throws Exception {
        server.enqueue(new MockResponse().setBody("<?xml version=\"1.0\"?><api><query><userinfo id=\"0\" name=\"foo\"/></query></api>"));

        boolean result = testObject.validateLogin();

        RecordedRequest loginTokenRequest = assertBasicRequestParameters(server, "GET");
        Map<String, String> params = parseQueryParams(loginTokenRequest);
        assertEquals("xml", params.get("format"));
        assertEquals("query", params.get("action"));
        assertEquals("userinfo", params.get("meta"));

        assertFalse(result);
    }

    @Test
    public void editToken() throws Exception {
        server.enqueue(new MockResponse().setBody("<?xml version=\"1.0\"?><api><tokens edittoken=\"baz\" /></api>"));

        String result = testObject.getEditToken();

        RecordedRequest loginTokenRequest = assertBasicRequestParameters(server, "GET");
        Map<String, String> params = parseQueryParams(loginTokenRequest);
        assertEquals("xml", params.get("format"));
        assertEquals("tokens", params.get("action"));
        assertEquals("edit", params.get("type"));

        assertEquals("baz", result);
    }

    @Test
    public void fileExistsWithName_FileNotFound() throws Exception {
        server.enqueue(new MockResponse().setBody("<?xml version=\"1.0\"?><api batchcomplete=\"\"><query> <normalized><n from=\"File:foo\" to=\"File:Foo\" /></normalized><pages><page _idx=\"-1\" ns=\"6\" title=\"File:Foo\" missing=\"\" imagerepository=\"\" /></pages></query></api>"));

        boolean result = testObject.fileExistsWithName("foo");

        RecordedRequest request = assertBasicRequestParameters(server, "GET");
        Map<String, String> params = parseQueryParams(request);
        assertEquals("xml", params.get("format"));
        assertEquals("query", params.get("action"));
        assertEquals("imageinfo", params.get("prop"));
        assertEquals("File:foo", params.get("titles"));

        assertFalse(result);
    }

    @Test
    public void getUploadCount() throws InterruptedException {
        server.enqueue(new MockResponse().setBody("23\n"));

        TestObserver<Integer> testObserver = testObject.getUploadCount("testUsername").test();

        RecordedRequest request = server.takeRequest();
        Map<String, String> params = parseQueryParams(request);
        assertEquals("testUsername", params.get("user"));

        assertEquals(1, testObserver.valueCount());
        assertEquals(23, (int)testObserver.values().get(0));
    }

    private RecordedRequest assertBasicRequestParameters(MockWebServer server, String method) throws InterruptedException {
        RecordedRequest request = server.takeRequest();
        assertEquals("/", request.getRequestUrl().encodedPath());
        assertEquals(method, request.getMethod());
        assertEquals("Commons/" + BuildConfig.VERSION_NAME + " (https://mediawiki.org/wiki/Apps/Commons) Android/" + Build.VERSION.RELEASE, request.getHeader("User-Agent"));
        if ("POST".equals(method)) {
            assertEquals("application/x-www-form-urlencoded", request.getHeader("Content-Type"));
        }
        return request;
    }

    private Map<String, String> parseQueryParams(RecordedRequest request) {
        Map<String, String> result = new HashMap<>();
        HttpUrl url = request.getRequestUrl();
        Set<String> params = url.queryParameterNames();
        for (String name : params) {
            result.put(name, url.queryParameter(name));
        }
        return result;
    }

    private Map<String, String> parseBody(String body) throws UnsupportedEncodingException {
        String[] props = body.split("&");
        Map<String, String> result = new HashMap<>();
        for (String prop : props) {
            String[] pair = prop.split("=");
            result.put(pair[0], URLDecoder.decode(pair[1], "utf-8"));
        }
        return result;
    }
}

File: app/src/test/kotlin/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApiTest.kt
package fr.free.nrw.commons.mwapi

import android.os.Build
import androidx.test.core.app.ApplicationProvider
import fr.free.nrw.commons.TestCommonsApplication
import fr.free.nrw.commons.kvstore.JsonKvStore
import fr.free.nrw.commons.utils.ConfigUtils
import okhttp3.OkHttpClient
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import okhttp3.mockwebserver.RecordedRequest
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.mock
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import org.wikipedia.util.DateUtil
import java.net.URLDecoder
import java.util.*

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [21], application = TestCommonsApplication::class)
class ApacheHttpClientMediaWikiApiTest {

    private lateinit var testObject: ApacheHttpClientMediaWikiApi
    private lateinit var server: MockWebServer
    private lateinit var wikidataServer: MockWebServer
    private lateinit var sharedPreferences: JsonKvStore
    private lateinit var okHttpClient: OkHttpClient

    @Before
    fun setUp() {
        server = MockWebServer()
        wikidataServer = MockWebServer()
        okHttpClient = OkHttpClient()
        sharedPreferences = mock(JsonKvStore::class.java)
        testObject = ApacheHttpClientMediaWikiApi("http://" + server.hostName + ":" + server.port + "/")
    }

    @After
    fun teardown() {
        server.shutdown()
    }

    @Test
    fun isUserBlockedFromCommonsForInfinitelyBlockedUser() {
        server.enqueue(MockResponse().setBody("<?xml version=\"1.0\"?><api><query><userinfo id=\"1000\" name=\"testusername\" blockid=\"3000\" blockedby=\"blockerusername\" blockedbyid=\"1001\" blockreason=\"testing\" blockedtimestamp=\"2018-05-24T15:32:09Z\" blockexpiry=\"infinite\"></userinfo></query></api>"))

        val result = testObject.isUserBlockedFromCommons()

        assertBasicRequestParameters(server, "GET").let { userBlockedRequest ->
            parseQueryParams(userBlockedRequest).let { body ->
                assertEquals("xml", body["format"])
                assertEquals("query", body["action"])
                assertEquals("userinfo", body["meta"])
                assertEquals("blockinfo", body["uiprop"])
            }
        }

        assertTrue(result)
    }

    @Test
    fun isUserBlockedFromCommonsForTimeBlockedUser() {
        val currentDate = Date()
        val expiredDate = Date(currentDate.time + 10000)
        server.enqueue(MockResponse().setBody("<?xml version=\"1.0\"?><api><query><userinfo id=\"1000\" name=\"testusername\" blockid=\"3000\" blockedby=\"blockerusername\" blockedbyid=\"1001\" blockreason=\"testing\" blockedtimestamp=\"2018-05-24T15:32:09Z\" blockexpiry=\"" + DateUtil.iso8601DateFormat(expiredDate) + "\"></userinfo></query></api>"))

        val result = testObject.isUserBlockedFromCommons()

        assertBasicRequestParameters(server, "GET").let { userBlockedRequest ->
            parseQueryParams(userBlockedRequest).let { body ->
                assertEquals("xml", body["format"])
                assertEquals("query", body["action"])
                assertEquals("userinfo", body["meta"])
                assertEquals("blockinfo", body["uiprop"])
            }
        }

        assertTrue(result)
    }

    @Test
    fun isUserBlockedFromCommonsForExpiredBlockedUser() {
        val currentDate = Date()
        val expiredDate = Date(currentDate.time - 10000)
        server.enqueue(MockResponse().setBody("<?xml version=\"1.0\"?><api><query><userinfo id=\"1000\" name=\"testusername\" blockid=\"3000\" blockedby=\"blockerusername\" blockedbyid=\"1001\" blockreason=\"testing\" blockedtimestamp=\"2018-05-24T15:32:09Z\" blockexpiry=\"" + DateUtil.iso8601DateFormat(expiredDate) + "\"></userinfo></query></api>"))

        val result = testObject.isUserBlockedFromCommons()

        assertBasicRequestParameters(server, "GET").let { userBlockedRequest ->
            parseQueryParams(userBlockedRequest).let { body ->
                assertEquals("xml", body["format"])
                assertEquals("query", body["action"])
                assertEquals("userinfo", body["meta"])
                assertEquals("blockinfo", body["uiprop"])
            }
        }

        assertFalse(result)
    }

    @Test
    fun isUserBlockedFromCommonsForNotBlockedUser() {
        server.enqueue(MockResponse().setBody("<?xml version=\"1.0\"?><api><query><userinfo id=\"1000\" name=\"testusername\"></userinfo></query></api>"))

        val result = testObject.isUserBlockedFromCommons()

        assertBasicRequestParameters(server, "GET").let { userBlockedRequest ->
            parseQueryParams(userBlockedRequest).let { body ->
                assertEquals("xml", body["format"])
                assertEquals("query", body["action"])
                assertEquals("userinfo", body["meta"])
                assertEquals("blockinfo", body["uiprop"])
            }
        }

        assertFalse(result)
    }

    private fun assertBasicRequestParameters(server: MockWebServer, method: String): RecordedRequest = server.takeRequest().let {
        assertEquals("/", it.requestUrl.encodedPath())
        assertEquals(method, it.method)
        assertEquals("Commons/${ConfigUtils.getVersionNameWithSha(ApplicationProvider.getApplicationContext())} (https://mediawiki.org/wiki/Apps/Commons) Android/${Build.VERSION.RELEASE}",
                it.getHeader("User-Agent"))
        if ("POST" == method) {
            assertEquals("application/x-www-form-urlencoded", it.getHeader("Content-Type"))
        }
        return it
    }

    private fun parseQueryParams(request: RecordedRequest) = HashMap<String, String?>().apply {
        request.requestUrl.let {
            it.queryParameterNames().forEach { name -> put(name, it.queryParameter(name)) }
        }
    }

    private fun parseBody(body: String): Map<String, String> = HashMap<String, String>().apply {
        body.split("&".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray().forEach { prop ->
            val pair = prop.split("=".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            put(pair[0], URLDecoder.decode(pair[1], "utf-8"))
        }
    }
}
